control/discovery.py

```diff
@@ -0,0 +1,958 @@
+NVME_TCP_ICRESP = 0x1
+NVME_TCP_H2C_TERM = 0x2
+NVME_TCP_C2H_TERM = 0x3
+NVME_TCP_CMD = 0x4
+NVME_TCP_RSP = 0x5
+NVME_TCP_H2C_DATA = 0x6
+NVME_TCP_C2H_DATA = 0x7
+NVME_TCP_R2T = 0x9
+
+# NVMe tcp opcode
+NVME_AQ_OPC_DELETE_SQ = 0x0
+NVME_AQ_OPC_CREATE_SQ = 0x1
+NVME_AQ_OPC_GET_LOG_PAGE = 0x2
+NVME_AQ_OPC_DELETE_CQ = 0x4
+NVME_AQ_OPC_CREATE_CQ = 0x5
+NVME_AQ_OPC_IDENTIFY = 0x6
+NVME_AQ_OPC_ABORT = 0x8
+NVME_AQ_OPC_SET_FEATURES = 0x9
+NVME_AQ_OPC_GET_FEATURES = 0xa
+NVME_AQ_OPC_ASYNC_EVE_REQ = 0xc
+NVME_AQ_OPC_NS_MGMT = 0xd
+NVME_AQ_OPC_FW_COMMIT = 0x10
+NVME_AQ_OPC_FW_IMG_DOWNLOAD = 0x11
+NVME_AQ_OPC_NS_ATTACH = 0x15
+NVME_AQ_OPC_KEEP_ALIVE = 0x18
+
+# NVMe tcp fabric command (special tcp opcode)
+NVME_FABRIC_OPC = 0x7F
+
+# NVMe tcp fabric command type
+NVME_FCTYPE_PROP_SET = 0x0
+NVME_FCTYPE_CONNECT = 0x1
+NVME_FCTYPE_PROP_GET = 0x4
+NVME_FCTYPE_AUTH_SEND = 0x5
+NVME_FCTYPE_AUTH_RECV = 0x6
+NVME_FCTYPE_DISCONNECT = 0x8
+
+# NVMe controller register space offsets
+NVME_CTL_CAPABILITIES = 0x0
+NVME_CTL_VERSION = 0x08
+NVME_CTL_CONFIGURATION = 0x14
+NVME_CTL_STATUS = 0x1c
+
+
+# NVM subsystem types
+# Discovery type for NVM subsystem
+NVMF_SUBTYPE_DISCOVERY = 0x1
+# NVMe type for NVM subsystem
+NVMF_SUBTYPE_NVME = 0x2
+
+# NVMe over Fabrics transport types
+NVMF_TRTYPE_RDMA = 0x1
+NVMF_TRTYPE_FC = 0x2
+NVMF_TRTYPE_TCP  = 0x3
+# Intra-host transport (loopback)
+NVMF_TRTYPE_INTRA_HOST = 0xfe
+
+# Address family types
+# IPv4 (AF_INET)
+NVMF_ADRFAM_IPV4 = 0x1
+# IPv6 (AF_INET6)
+NVMF_ADRFAM_IPV6 = 0x2
+# InfiniBand (AF_IB)
+NVMF_ADRFAM_IB = 0x3
+# Fibre Channel address family
+NVMF_ADRFAM_FC = 0x4
+# Intra-host transport (loopback)
+NVMF_ADRFAM_INTRA_HOST	= 0xfe
+
+# Transport requirement, secure channel requirements
+# Connections shall be made over a fabric secure channel
+NVMF_TREQ_SECURE_CHANNEL_NOT_SPECIFIED = 0x0
+NVMF_TREQ_SECURE_CHANNEL_REQUIRED = 0x1
+NVMF_TREQ_SECURE_CHANNEL_NOT_REQUIRED = 0x2
+
+# NVMe tcp package length, refer: MTU = 1500 bytes
+NVME_TCP_PDU_UNIT = 1024
+
+# Max SQ head pointer
+SQ_HEAD_MAX = 128
+
+lock = threading.Lock()
+
+# Global controller id
+GLOBAL_CNLID = 0x1
+
+# Global generation counter
+GLOBAL_GEN_CNT = 0x1
+
+class Pdu:
+    def __init__(self):
+        # PDU type
+        self.type = bytearray(1)
+        # PDU specical flag
+        self.specical_flag = bytearray(1)
+        # PDU header length
+        self.header_length = bytearray(1)
+        # PDU data offset
+        self.data_offset = bytearray(1)
+        # packet length
+        self.packet_length = bytearray(4)
+
+    def compose_reply(self):
+        return self.type + self.specical_flag + self.header_length + \
+               self.data_offset + self.packet_length
+
+class ICResp:
+    def __init__(self):
+        # pdu version format
+        self.version_format = bytearray(2)
+        # controller Pdu data alignment
+        self.data_alignment = bytearray(1)
+        # digest types enabled
+        self.digest_types = bytearray(1)
```

Comment:
I don't see any calculation of header and data digests nor digest checking and error handling.

Reply:
I don't quite understand. Do you mean there are many fields that have not been assigned values here? Because the default value of 0 was used. These classes are all built for replies and do not require error handling. Some received messages did not have validation and error handling done.